diff --git a/core/jni/android_os_GraphicsEnvironment.cpp b/core/jni/android_os_GraphicsEnvironment.cpp
index 905a85adc55..a8c86a93b32 100644
--- a/core/jni/android_os_GraphicsEnvironment.cpp
+++ b/core/jni/android_os_GraphicsEnvironment.cpp
@@ -15,8 +15,9 @@
  */
 
 #define LOG_TAG "GraphicsEnvironment"
-
+#ifndef MTK_HARDWARE
 #include <gui/GraphicsEnv.h>
+#endif
 #include <nativehelper/ScopedUtfChars.h>
 #include "core_jni_helpers.h"
 
@@ -24,7 +25,9 @@ namespace {
 
 void setDriverPath(JNIEnv* env, jobject clazz, jstring path) {
     ScopedUtfChars pathChars(env, path);
+#ifndef MTK_HARDWARE
     android::GraphicsEnv::getInstance().setDriverPath(pathChars.c_str());
+#endif
 }
 
 const JNINativeMethod g_methods[] = {
diff --git a/core/jni/fd_utils-inl.h b/core/jni/fd_utils-inl.h
index a68a60274ac..c6fc5088577 100644
--- a/core/jni/fd_utils-inl.h
+++ b/core/jni/fd_utils-inl.h
@@ -58,6 +58,7 @@ static const char* kPathWhitelist[] = {
   "/dev/urandom",
   "/dev/ion",
   "/dev/dri/renderD129", // Fixes b/31172436
+  "/proc/ged", // MTK add
   "/system/framework/org.cyanogenmod.platform-res.apk",
 #ifdef PATH_WHITELIST_EXTRA_H
 PATH_WHITELIST_EXTRA_H
